### PR TITLE
Registry: Accept staking nodes even if they are not enrolled

### PR DESCRIPTION
**Pull request by Geod24**
One of the issue we are seeing live is that non-enrolled validators cannot yet register, leading to jumpy DNS.
In order to avoid this issue, we should rely on the stake being present, as it's much more long-lived.

### CODE REVIEW

**Review by mkykadir — APPROVED** [review]

File: source/agora/consensus/state/Ledger.d [hunk 1]
```diff
@@ -587,6 +587,19 @@ public class Ledger
         }
     }
 
+    /// Returns: Whether the `utxo` can be used as a stake for a validator
+    public bool isStake (in Hash hash, in Output utxo) scope @safe
```
Comment (omerfirmak) [inline source/agora/consensus/state/Ledger.d:591]: `getPenaltyDeposit() > 0` would be equivalent

**Review by omerfirmak — APPROVED** [review]

File: source/agora/consensus/state/Ledger.d (hunk 1 above)
Comment (Geod24) [reply source/agora/consensus/state/Ledger.d:591]: I have questions about that name... It's what it seems to be doing but it is not at all clear what other role(s) it takes ?
Also this doesn't do the UTXO lookup.

File: source/agora/consensus/state/Ledger.d [hunk 2]
```diff
@@ -1027,3 +1040,80 @@ unittest
...
+{
+    import agora.common.Set;
+    import std.typecons;
+
+    /// Type of data we keep track of
+    public struct Tracked
+    {
+        /// Hash of the UTXO
+        public Hash hash;
+        /// Output itself
```
Comment (linked0) [inline source/agora/consensus/state/Ledger.d:1064]: Need a line break.

File: source/agora/consensus/state/Ledger.d [hunk 3]
```diff
@@ -1027,3 +1040,80 @@ unittest
...
+        tx.outputs.enumerate
+            .map!((in tup) => tuple(UTXO.getHash(txHash, tup.index), tup.value))
+            .each!((in hash, in utxo) => this.externalize(hash, utxo));
+    }
+
+    /// Process a single `Output`
+    public void externalize (in Hash hash, in Output output) @safe
+    {
+        if (this.include(hash, output))
+            this.data_.put(Tracked(hash, output));
```
Comment (linked0) [inline source/agora/consensus/state/Ledger.d:1118]: Isn't there any possibility that the `data_` would be ever-growing to some critical point?

File: source/agora/consensus/state/Ledger.d (hunk 1 above)
Comment (omerfirmak) [reply source/agora/consensus/state/Ledger.d:591]: It has a history:) At first it was called `hasPenaltyDeposit` and returned a `bool`. But after I decided to let the melting TX spend the deposit, I needed a way to get how much the deposit was as well. Deposit is fixed and configured in `ConsensusParams` but our TX validation routine have no access to it.

File: source/agora/consensus/state/Ledger.d (hunk 2 above)
Comment (Geod24) [reply source/agora/consensus/state/Ledger.d:1064]: Fixed

File: source/agora/consensus/state/Ledger.d (hunk 3 above)
Comment (Geod24) [reply source/agora/consensus/state/Ledger.d:1118]: It's more likely that `UTXOSet` reach that threshold way before this struct does, as this is only a (likely small) subset of `UTXOSet`.

Comment (linked0) [reply source/agora/consensus/state/Ledger.d:1118]: Right. But `updateUTXOCache` has the `remove` operation but it doesn't. So I was just a little worried about that.

**Review by linked0 — APPROVED** [review]
LGTM

File: source/agora/consensus/state/Ledger.d (hunk 3 above)
Comment (Geod24) [reply source/agora/consensus/state/Ledger.d:1118]: See line 1107

**Pull request merged**